OutOfSchool/OutOfSchool.Redis/MultiLayerCache.cs

```diff
@@ -0,0 +1,125 @@
+﻿using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.Options;
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace OutOfSchool.Redis;
+
+public sealed class MultiLayerCache : IMultiLayerCacheService, IDisposable
+{
+    private readonly ReaderWriterLockSlim cacheLock = new ReaderWriterLockSlim();
```

Comment:
Could you explain what purpose this `cacheLock` serves? 

The MemoryCache is thread safe (see [MS doc](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.caching.memorycache?view=dotnet-plat-ext-7.0#thread-safety)).

Current implementation makes all writes **blocking** each other, so changing cache items with **diferent keys** is done **sequentially**. Is it expected? What is the trade to **loose performance**?

Reply:
I read about it. But all samples, that I found, were with lockers. And I used the same approach.

Reply:
"Quotes found on the Internet are not always accurate" (c) Lao Tzu & (later) Abraham Lincoln

(More seriously) Could you share links to that samples?

Reply:
https://habr.com/ru/companies/otus/articles/502974/
https://code-maze.com/aspnetcore-in-memory-caching/
https://michaelscodingspot.com/cache-implementations-in-csharp-net/

And some words about MemoryCache in this article:
https://blog.novanet.no/asp-net-core-memory-cache-is-get-or-create-thread-safe/

Reply:
Thank you for links. It was interesting reading, especially the github code of LazyCache mentioned in last article.

At same time it doesn't answer any of my questions regarding your implementation. (TLDR: in articles I see Liverpool, in your code I see Manchester - in general both are about english football, both are about different things.)

---

But let's read them together, step-by-step. 
- 1st link "Habr" - suggests to use locks per each cache key (`ConcurrentDictionary<object, SemaphoreSlim>`). Not 1 lock for all. At same time comments explains why it is bad. Noone is suggesting to use 1 lock for all. 
(Btw, has to mention good example of `if-lock-if` pattern usage in last implementation of `WaitToFinishMemoryCache`. It's good to understand why it is better than simple `if-lock` and use it wherever possible. I.e. check your implementation of `SetAsync<T>()`. Ofc it doesn't mean, that the whole idea of `_locks` is good.)
- 2nd link "Code-Maze" - single `SemaphoreSlim` is used for isolated case of GetOrCreate. Not everywhere. Thus blocking issue is bit reduced, but still same - two parallel cache additions would block cache even when it is done for different keys.
- 3rd link "Michael's coding spot" - (it has slightly broken code, but I hope I got it) again `ConcurrentDictionary<object, SemaphoreSlim>` is used. And no surprise, b/c Habr article was translation of this one. So all issues still are still the same.
- 4th link "nova net" - mentioned same issue with isolated `GetOrCreate`, not all cache methods. As a good cache article suggests 3rd-party library - LazyCache and inside implementation [github: CachingService](https://github.com/alastairtree/LazyCache/blob/master/LazyCache/CachingService.cs#L103) you could see that it's not so easy as your implementation.

---

Summing all up. 
- Issues mentioned in my first comment are not solved in neither article except last one. 
- Code, suggested in all articles targets only one cache method `GetOrCreate`/`GetOrAdd`, not all cache methods, and "improved versions" of that code suggests to use as many locks as cache keys (or as count of hashes of cache key - for LazyCache [see lines 119-120](https://github.com/alastairtree/LazyCache/blob/master/LazyCache/CachingService.cs#L119)). 
- Please take all things written in Internet with a grain of salt, especially, if there is no comment section. 
- If you feel that you are not convinced - please call me in Teams, we could discuss it in voice.
- If you don't want to change anything and sure that your code is perfect - please state it here. I'll stop asking for improvements and will close my comments as "resolved".

Reply:
The other small point to add to the conversation. I see it now, that CacheService is using the same locking mechanism with different lock. Hadn't analyzed it thoroughly but there might be potential for dead-locks.

Reply:
My code is not perfect ))))
I like your comments and I always wait for them.
They help me improve my skills and understanding.
After reading your comments I can look at my code from another side.
And I am very grateful for the time that you spend reviewing and analyzing my PRs.

I try to change something.

Reply:
Deleted locker